File: benchmarks/solutions/haskell/a3-valid-parens.hs
module ValidParens where

isValid :: String -> Bool
isValid = go []
  where
    go [] []     = True
    go _  []     = False
    go stack (c:cs)
      | c `elem` "([{" = go (c:stack) cs
      | otherwise = case stack of
          []     -> False
          (x:xs) -> matching x c && go xs cs
    matching '(' ')' = True
    matching '[' ']' = True
    matching '{' '}' = True
    matching _   _   = False

File: benchmarks/solutions/go/a3-valid-parens.go
package main

func isValid(s string) bool {
	stack := []rune{}
	pairs := map[rune]rune{')': '(', ']': '[', '}': '{'}
	for _, c := range s {
		if open, ok := pairs[c]; ok {
			if len(stack) == 0 || stack[len(stack)-1] != open {
				return false
			}
			stack = stack[:len(stack)-1]
		} else {
			stack = append(stack, c)
		}
	}
	return len(stack) == 0
}
